dispatchers: warn user if try to use EXPOSE ip:hostPort:containerPort

chenhanxiao opened the pull request:
We could use EXPOSE ip:hostPort:containerPort,
but actually it did as EXPOSE ::containerPort

commit 2275c833 already warned user on daemon side.
This patch will print warning message on client side.

Signed-off-by: Chen Hanxiao chenhanxiao@cn.fujitsu.com

thaJeztah commented:
Replaces https://github.com/docker/docker/pull/10629 and https://github.com/docker/docker/pull/10630?

chenhanxiao commented:
@thaJeztah  @duglin 
I closed #10630  by accident(by a chaos force push), sorry for that.

I updated the testcases as your comments, it works fine now.

duglin commented:
Did you ever figure out why the test case wasn't showing you the output? 

-Doug

Sent from my iPhone

> On Feb 15, 2015, at 6:33 AM, Chen Hanxiao <notifications@github.com> wrote:
> 
> @thaJeztah @duglin 
> I closed #10630 by accident(by a chaos force push), sorry for that.
> 
> I updated the testcases as your comments, it works fine now.
> 
> —
> Reply to this email directly or view it on GitHub.
>

chenhanxiao commented:
@duglin 

It failed with the same codes on my local tree last whole weeks.

Today I found Jenkins could not automerge my branch,
so I rebase it on upstream's master, rebuild it, then it works fine...

thaJeztah commented:
@chenhanxiao alright then, accidents happen :). Speaking for myself here, but perhaps you could add a short note to the old PRs, referring to the new ones (e.g., *"closed this PR because of merge errors, replaced by xxxx"*). There are a *lot* of PRs in this repository and it can become hard to find out *why* a PR was closed and *where* it is continued.

duglin commented:
LGTM
ping @tiborvass @erikh

chenhanxiao commented:
@thaJeztah

Sure, I'll add a comment at #10630.
Thanks for your advice.

chenhanxiao commented:
@tiborvass @erikh
ping?

LK4D4 commented on builder/dispatchers.go:
```diff
@@ -339,11 +339,19 @@ func expose(b *Builder, args []string, attributes map[string]bool, original stri
 
+	for _, bindings := range bindingMap {
+		if bindings[0].HostIp != "" || bindings[0].HostPort != "" {
+			fmt.Fprintf(b.OutStream, " ---> The mapping to public ports on your host via:\n"+
```
Are you sure that this shouldn't be in `b.ErrStream`?

tiborvass replied:
@chenhanxiao would you mind fixing this? if not, let us know, we can carry your patch.

chenhanxiao replied:
@tiborvass
Sorry for the late reply as I'm on travel in the past few days.

I'll update this today.

chenhanxiao replied:
@LK4D4 
It should be b.ErrStream, sorry for my carelessness.

thaJeztah commented:
/ping @tiborvass I see the PR is updated

chenhanxiao commented:
@tiborvass @erikh

ping?

tiborvass commented:
LGTM

Ping @duglin

tiborvass commented:
Ping @moxiegirl @SvenDowideit @fredlf  @jamtur01

jamtur01 commented:
Only concern I have is that people link this to mapping with EXPOSE as opposed to EXPOSE being deprecated? Otherwise LGTM.

tiborvass commented:
@jamtur01 `EXPOSE` is not deprecated, only a special archaic form of it.

jamtur01 commented:
@tiborvass That's exactly what I mean - it's not 100% clear from the error message.

tiborvass commented:
@jamtur01 oh I see. Well that's why I wanted you guys to proof read it and potentially provide a less ambiguous formulation :)

jamtur01 commented on builder/dispatchers.go:
```diff
@@ -339,11 +339,19 @@ func expose(b *Builder, args []string, attributes map[string]bool, original stri
+		if bindings[0].HostIp != "" || bindings[0].HostPort != "" {
+			fmt.Fprintf(b.ErrStream, " ---> The mapping to public ports on your host via:\n"+
+				"      Dockerfile EXPOSE (ip:hostPort:containerPort) has been deprecated.\n"+
+				"      Use -p to publish the ports.\n")
```
I think it's the interaction of mapping and publish that's throwing me. Perhaps?

Using Dockerfile's EXPOSE instruction to map host ports to container ports is deprecated. Please use the -p flag to map ports.

chenhanxiao replied:
@jamtur01 
I think your version is clearer.
@SvenDowideit 
I want to use:
Using Dockerfile's EXPOSE instruction to map host ports to container ports(ip:hostPort:containerPort) is deprecated. Please use the -p flag to map ports.

How do you think?

SvenDowideit replied:
yup, sounds good to me

SvenDowideit commented on builder/dispatchers.go:
```diff
@@ -339,11 +339,19 @@ func expose(b *Builder, args []string, attributes map[string]bool, original stri
+			fmt.Fprintf(b.ErrStream, " ---> The mapping to public ports on your host via:\n"+
+				"      Dockerfile EXPOSE (ip:hostPort:containerPort) has been deprecated.\n"+
+				"      Use -p to publish the ports.\n")
+		}
```
can we use one phrase - we're using 2 here `mapping to public ports` and `publish the ports`

chenhanxiao committed:
dispatchers: warn user if try to use EXPOSE ip:hostPort:containerPort

We could use EXPOSE ip:hostPort:containerPort,
but actually it did as EXPOSE ::containerPort

commit 2275c833 already warned user on daemon side.
This patch will print warning message on client side.

Signed-off-by: Chen Hanxiao <chenhanxiao@cn.fujitsu.com>

chenhanxiao commented:
@SvenDowideit @fredlf @jamtur01

PR updated as your comments.

Regards,
- Chen

tiborvass commented:
Ping @SvenDowideit @moxiegirl  @fredlf @jamtur01

moxiegirl commented:
LGTM
